### PR TITLE
Turns the baseturf var into a part time list

### LINKED ISSUE #16391: Placing floor tiles on asteroid plating has strange interactions
Using generic plasteel tiles.
1. Place a tile. It will be invisible/look like the default asteroid turf with no variants
2. Pry up the tile you just placed. The asteroid will now look like metal plating.

### LINKED ISSUE #32776: Burning a wall with thermite also destorys the floor beneath.
Title, they are supposed to leave burnt-out floors but instead make a hole in space.

### PR DESCRIPTION
**Pull request by ninjanomnom**
:cl: ninjanomnom
fix: Things like thermite which burned through walls straight to space now stop on plating. You'll have to thermite it again to get to space.
/:cl:

Changes the baseturf var into a list where it's needed. Most of the time it'll just be a single type like before but the ScrapeAway(), PlaceOnTop(), PlaceOnBottom(), and CopyOnTop() procs handle turning it to and from a list.

The baseturfs list is generated in initialization and type definitions should never be a list by default or things will break. If you're worried about the performance cost of this:
image
It only generates the list for a given baseturf starting point once. Every further call just copies the pre-existing result.

This allows us to have multiple layers of turfs before digging down further doesn't change the turf type and allows shuttles (eventually) to be able to keep track of more turfs under it, thus allowing shuttles to land on more complex landing zones than single types.

Shuttles and thermite work as intended, something is going to be broken though so testmerging is required.

This also does some minor code improvements to the ChangeTurf family of procs and moves them to their own file.

fixes #32776
fixes #16391 (The second part)

### CODE REVIEW

**Comment by Buggy123** [comment]
_Theoretically speaking_, this means that you could, say, allow players to emag navigation consoles, and then land *in* the station, right?

**Comment by ninjanomnom** [comment]
Yes though in the case of a shuttle landing inside the station it'd make more sense if it broke a hole.

If you want it to keep all the turfs that were there before though sure it works

**Comment by Xhuis** [comment]
With luck, after he shows he's a good worker, we can take Baseturf on as a full-time list, too.

File: _maps/RandomRuins/LavaRuins/lavaland_surface_pizzaparty.dmm [hunk 1]
```diff
@@ -11,21 +11,21 @@
 /area/lavaland/surface/outdoors)
 "d" = (
 /turf/closed/wall{
-	baseturf = /turf/open/floor/plating/asteroid/basalt/lava_land_surface
+	baseturfs = /turf/open/floor/plating/asteroid/basalt/lava_land_surface
```
Comment (ShizCalev) [inline _maps/RandomRuins/LavaRuins/lavaland_surface_pizzaparty.dmm]: Remove these, there's a baseturf helper down.

File: _maps/RandomRuins/SpaceRuins/deepstorage.dmm [hunk 2]
```diff
@@ -20,32 +20,32 @@
 	})
 "af" = (
 /turf/closed/wall/mineral/iron{
-	baseturf = /turf/open/floor/plating/asteroid/airless
+	baseturfs = /turf/open/floor/plating/asteroid/airless
```
Comment (ShizCalev) [inline _maps/RandomRuins/SpaceRuins/deepstorage.dmm]: Replace with baseturf helpers.

**Comment by arsserpentarium** [comment]
So that's way to implement archeology?

File: _maps/RandomRuins/SpaceRuins/spacehotel.dmm [hunk 3]
```diff
@@ -4060,12 +4060,12 @@
 "li" = (
 /obj/effect/light_emitter,
 /turf/open/floor/plating/beach/water{
-	baseturf = /turf/open/space
+	baseturfs = /turf/open/space
```
Comment (ShizCalev) [inline _maps/RandomRuins/SpaceRuins/spacehotel.dmm]: Replace with a space baseturf helper.

File: _maps/RandomZLevels/beach2.dmm [hunk 4]
```diff
@@ -92,20 +92,20 @@
 	pixel_y = 18
 	},
 /turf/open/floor/wood{
-	baseturf = /turf/open/floor/plating/beach/sand
+	baseturfs = /turf/open/floor/plating/beach/sand
```
Comment (ShizCalev) [inline _maps/RandomZLevels/beach2.dmm]: Delete and replace with a beach sand baseturf helper.

File: _maps/RandomZLevels/caves.dmm [hunk 5]
```diff
@@ -10,7 +10,7 @@
 /area/awaymission/caves/BMP_asteroid/level_three)
 "ad" = (
 /turf/open/lava/smooth{
-	baseturf = /turf/open/lava/smooth;
+	baseturfs = /turf/open/lava/smooth;
```
Comment (ShizCalev) [inline _maps/RandomZLevels/caves.dmm]: Replace with a lava baseturf helper

File: _maps/RandomZLevels/snowdin.dmm [hunk 6]
```diff
@@ -10,14 +10,14 @@
 /area/awaymission/snowdin)
 "ad" = (
 /turf/closed/wall/ice{
-	baseturf = /turf/open/floor/plating/asteroid/snow
+	baseturfs = /turf/open/floor/plating/asteroid/snow
```
Comment (ShizCalev) [inline _maps/RandomZLevels/snowdin.dmm]: snow baseturf helper

File: _maps/shuttles/emergency_delta.dmm [hunk 7]
```diff
@@ -353,7 +353,7 @@
 	name = "emergency shower"
 	},
 /turf/open/floor/plasteel/whiteblue/side{
-	baseturf = /turf/open/lava/smooth/lava_land_surface;
+	baseturfs = /turf/open/lava/smooth/lava_land_surface;
```
Comment (ShizCalev) [inline _maps/shuttles/emergency_delta.dmm]: This seems to be a mistake.

File: _maps/RandomZLevels/caves.dmm [hunk 8]
```diff
@@ -1748,20 +1748,20 @@
...
 /turf/open/floor/plasteel{
-	baseturf = /turf/open/floor/plating/asteroid/basalt
+	baseturfs = /turf/open/floor/plating/asteroid/basalt
 	},
 /area/awaymission/caves/BMP_asteroid)
 "eB" = (
 /obj/structure/barricade/wooden,
 /turf/open/floor/plasteel{
-	baseturf = /turf/open/floor/plating/asteroid/basalt
+	baseturfs = /turf/open/floor/plating/asteroid/basalt
```
Comment (ShizCalev) [inline _maps/RandomZLevels/caves.dmm]: basalt baseturf helper

**Review by ShizCalev — CHANGES_REQUESTED** [review]

**Comment by Buggy123** [comment]
This does mean you could do things like pressurized shuttle bays with a floor, which is neat.

**Comment by ninjanomnom** [comment]
That's already easily possible

**Comment by ninjanomnom** [comment]
@ShizCalev Map baseturf helpers added and old baseturf var edits removed with these two regexes:
`\sbaseturfs *= *.+?;\n`
`{\n\sbaseturfs *= *.+?\n\s}`

**Review by ShizCalev — DISMISSED** [review]
Mapping.

File: code/game/turfs/turf.dm [hunk 9]
```diff
@@ -180,6 +188,54 @@
...
+	if(fake_baseturf_type)
+		if(length(fake_baseturf_type)) // We were given a list, just apply it and move on
+			baseturfs = fake_baseturf_type
+			return
+		current_target = fake_baseturf_type
+	else
+		if(length(baseturfs))
+			return // No replacement baseturf has been given and the current baseturfs value is already a list/assembled
+		if(!baseturfs)
+			current_target = initial(baseturfs) || type // This should never happen but just in case...
```
Comment (Cyberboss) [inline code/game/turfs/turf.dm]: maybe add a stack_trace() then?

**Review by Cyberboss — DISMISSED** [review]
Mapping module

**Comment by ninjanomnom** [comment]
This has been tested enough, we're good to go whenever I think

**Pull request merged**